Sources/BigInt/BigInt.swift

```diff
@@ -0,0 +1,799 @@
+        }
+        newWords[lastJ + 1] = digit
+      }
+    }
+
+    for i in stride(from: count - 1, through: 1, by: -1) {
+      if newWords[i] == 0, newWords[i - 1] <= Int.max {
+        newWords.removeLast()
+      } else {
+        break
+      }
+    }
+
+    if lhsIsNeg || rhsIsNeg, !(lhsIsNeg && rhsIsNeg) {
+      return -BigInt(_uncheckedWords: newWords)
+    }
+
+    return BigInt(_uncheckedWords: newWords)
+  }
+
+  @inlinable
+  public static func *= (lhs: inout BigInt, rhs: BigInt) {
+    lhs = lhs * rhs
+  }
+}
+
+extension BigInt: SignedNumeric {
+
+  public mutating func negate() {
+    var isOverflow = true
+    for i in 0 ..< words.count {
+      if isOverflow {
+        (words[i], isOverflow) = (~words[i]).addingReportingOverflow(1)
+      } else {
+        words[i] = ~words[i]
+      }
+    }
+
+    BigInt._dropExcessWords(words: &words)
+  }
+
+  @inlinable
+  public static prefix func - (x: BigInt) -> BigInt {
+    var result = x
+    result.negate()
+    return result
+  }
+}
+
+extension BigInt: BinaryInteger {
+
+  public init?<T>(exactly source: T) where T: BinaryFloatingPoint {
+    guard source.isFinite, source == source.rounded(.towardZero) else {
+      return nil
+    }
+    self.init(source)
+  }
+
+  public init<T>(_ source: T) where T: BinaryFloatingPoint {
+    precondition(
+      source.isFinite,
+      """
+      \(type(of: source)) value cannot be converted to BigInt because it is \
+      either infinite or NaN
+      """)
+
+    let isNegative = source < 0.0
+    var float = isNegative ? -source : source
+
+    if let _ = UInt(exactly: T.greatestFiniteMagnitude) {
+      words = [UInt(float)]
+    } else {
+      var words = Words()
+      let radix = T(sign: .plus, exponent: T.Exponent(UInt.bitWidth), significand: 1)
+      repeat {
+        let digit = UInt(float.truncatingRemainder(dividingBy: radix))
+        words.append(digit)
+        float = (float / radix).rounded(.towardZero)
+      } while float != 0
+
+      if let last = words.last, last >= Int.max {
+        words.append(0)
+      }
+      self.words = words
+    }
+
+    if isNegative {
+      self.negate()
+    }
+  }
+
+  public init<T>(_ source: T) where T: BinaryInteger {
+    if source >= 0, source < BigInt._digits.count {
+      self = BigInt._digits[Int(source)]
+    } else {
+      words = Words(source.words)
+      if source > 0 && source.words[source.words.endIndex - 1] > Int.max {
+        words.append(0)
+      }
+      // needed to handle sign-extended multi-word numbers that
+      // actually fit in a single word
+      BigInt._dropExcessWords(words: &words)
+    }
+  }
+
+  public init<T>(clamping source: T) where T: BinaryInteger {
+    self.init(source)
+  }
+
+  public init<T>(truncatingIfNeeded source: T) where T: BinaryInteger {
+    words = Words(source.words)
+  }
+
+  public var bitWidth: Int { words.count * UInt.bitWidth }
+
+  public var trailingZeroBitCount: Int { words.first?.trailingZeroBitCount ?? 0 }
```

Comment:
Shouldn’t that be something like
```Swift
var trailingZeroBitCount: Int {
    var totalZeros = 0
    for word in words {
        let zeros = word.trailingZeroBitCount
        totalZeros += zeros
        if zeros < UInt.bitWidth {
            break
        }
    }
    return totalZeros
}
```

Reply:
No, imagine that the entire array of `Uint`s is one giant integer. Only the trailing zeros in the least significant word are actually trailing, the rest are in the middle!

Reply:
If the least significant word is all zeroes (e.g. `UInt.max + 1`), you'll need to keep going up the chain.

Reply:
Doh! You're right! I feel suddenly stupid, I'll fix that.

Reply:
Fixed it and added some quick tests for it too.